### PR TITLE
OCPBUGS-35228: Fix desired before sync_worker's work is initialized

### PR DESCRIPTION
**Pull request by hongkailiu**
If the CVO pod gets restarted after a user sends a cluster
upgrade request with `--to-image`, the desired should not ALWAYS be
taken from `.Spec.DesiredUpdate`, e.g., when the targeting payload is blocked
on some precondition check failure.

The user request will be written to `cv.spec.desiredUpdate` where only
the `image` field is set, i.e., the `Version` field is empty.

If the CVO gets restarted, then

1. The desied [1] from `cv.spec.desiredUpdate` will be used to set
   `w.status.Actual` [2] and then applied to `cv.status.desired`.
2. When the next `optr.sync()` happens, the precondition check may become
   non-blocking [3] as the `Version` field is empty now.
3. CVO will then start an upgrade that should have been blocked.

[1]. https://github.com/openshift/cluster-version-operator/blob/1995380b6d755c29b926b846a64ca0039002c2cf/pkg/cvo/cvo.go#L680
[2]. https://github.com/openshift/cluster-version-operator/blob/1995380b6d755c29b926b846a64ca0039002c2cf/pkg/cvo/sync_worker.go#L502
[3]. https://github.com/openshift/cluster-version-operator/blob/1995380b6d755c29b926b846a64ca0039002c2cf/pkg/payload/precondition/clusterversion/rollback.go#L57

### CODE REVIEW

File: pkg/cvo/cvo_scenarios_test.go
Comment (wking) [inline pkg/cvo/cvo_scenarios_test.go:2964]: I'm not clear on where this `1.0.1-abc` vs. `1.0.0-abc` divergence comes from.  From what I understand, the test case is setting `image/image:0` up as `1.0.0-abc`, but this line is claiming it's `1.0.1-abc`?

Comment (hongkailiu) [reply pkg/cvo/cvo_scenarios_test.go:2964]: Did a little digging:
The line for the msg is:
https://github.com/openshift/cluster-version-operator/blob/f0139d19c34dbddd4a59148d472aa44d468660b3/pkg/cvo/sync_worker.go#L352

It can be traced back from the input of the case:
https://github.com/openshift/cluster-version-operator/pull/1081/files#diff-6cca1696ca6d83b25f5baf586b99602d32827c05c4276dd03ebfcb6825ff2558R2866
and eventually from
https://github.com/openshift/cluster-version-operator/blob/58356deee675c900b7cd4742ceefc0e6a84d7efb/pkg/cvo/testdata/payloadtest-2/release-manifests/release-metadata

The setup of this new case was duplicated from https://github.com/openshift/cluster-version-operator/blob/f0139d19c34dbddd4a59148d472aa44d468660b3/pkg/cvo/cvo_scenarios_test.go#L2859 with `worker.stillInitializingFunc = func() bool { return false }` removed which is for covering the introduced branch.

Comment (hongkailiu) [reply pkg/cvo/cvo_scenarios_test.go:2964]: Maybe https://github.com/openshift/cluster-version-operator/blob/58356deee675c900b7cd4742ceefc0e6a84d7efb/pkg/cvo/testdata/payloadtest/release-manifests/release-metadata makes more sense now?

Let me give it a try.

File: pkg/cvo/sync_worker.go
Comment (petr-muller) [inline pkg/cvo/sync_worker.go:238]: The only production code callsite calls the returned func right away, do we really need the indirection? I haven't looked at the tests yet but it seems to me we could simply return the bool here.

File: pkg/cvo/sync_worker.go
Comment (petr-muller) [inline pkg/cvo/sync_worker.go:195]: this exists only for tests, right? if so let's tell the reader in a comment

File: pkg/cvo/cvo.go
Comment (petr-muller) [inline pkg/cvo/cvo.go:681]: We use the name negated, can we call this `initialized` instead (the method too? To me it also seems that we don't get much value from the indirection as we call the returned func right away:

```
initialized := optr.configSync.Initialized()

if ok && initialized {
 // happy
...
```

File: pkg/cvo/cvo.go
Comment (petr-muller) [inline pkg/cvo/cvo.go:685]: Maybe call this `pendingDesired`? Something that we still need to evaluate?

File: pkg/cvo/cvo.go
Comment (petr-muller) [inline pkg/cvo/cvo.go:691]: "stillInitializing` (or `initialized` like I suggest above) carries more meaning than `ok`, perhaps we could use that here to convey the intent better (maybe we would even not need a comment about implied value of the other variable)

File: pkg/cvo/sync_worker.go
Comment (hongkailiu) [reply pkg/cvo/sync_worker.go]: It makes easier to fix the existing unit tests.

File: pkg/cvo/sync_worker.go
Comment (hongkailiu) [reply pkg/cvo/sync_worker.go]: Done!

File: pkg/cvo/cvo.go
Comment (hongkailiu) [reply pkg/cvo/cvo.go]: Renamed the function and variable.

File: pkg/cvo/cvo.go
Comment (hongkailiu) [reply pkg/cvo/cvo.go]: Done

File: pkg/cvo/cvo.go
Comment (hongkailiu) [reply pkg/cvo/cvo.go]: Updated.

**Comment by petr-muller** [comment]
@hongkailiu check out https://github.com/petr-muller/cluster-version-operator/commit/b8039aafdab8edcfa7de5181ad4e61b45d3924ce

File: pkg/cvo/sync_worker.go
Comment (Davoska) [inline pkg/cvo/sync_worker.go:38]: Nit: `ConfigSyncWorker` is an interface.

```suggestion
	// Initialized returns true if sync worker has work to do already
```

**Review by Davoska — CHANGES_REQUESTED** [review]
I have no objections regarding the fix, nice work :raised_hands: 

Only a few nitpicks regarding wording and unit tests, but nothing functional.

File: pkg/cvo/cvo.go
Comment (hongkailiu) [reply pkg/cvo/cvo.go]: Only made a difference from your suggestion.
sync worker -> worker

File: pkg/cvo/cvo_scenarios_test.go
Comment (hongkailiu) [reply pkg/cvo/cvo_scenarios_test.go]: Done.

File: pkg/cvo/cvo_scenarios_test.go
Comment (hongkailiu) [reply pkg/cvo/cvo_scenarios_test.go]: Removed.

File: pkg/cvo/sync_worker.go
Comment (hongkailiu) [reply pkg/cvo/sync_worker.go]: sync worker -> the worker

File: pkg/cvo/cvo.go
Comment (Davoska) [reply pkg/cvo/cvo.go]: Resolved, thanks.

File: pkg/cvo/sync_worker.go
Comment (Davoska) [reply pkg/cvo/sync_worker.go]: Thanks.

File: pkg/cvo/cvo_scenarios_test.go
Comment (hongkailiu) [reply pkg/cvo/cvo_scenarios_test.go]: For the existing test:
https://github.com/openshift/cluster-version-operator/pull/1088

**Comment by Davoska** [comment]
/approved

The changes look good; I am adding the approved label. As there were reviews from multiple people, I'll leave the lgtm label to one of the other reviewers to verify that the requested changes were made.

Before merging, let's squash the commits using the /label tide/merge-method-squash command or by manually rebasing.

Disclaimer: I have not tested the PR in a live cluster.

**Comment by Davoska** [comment]
/approve
:facepalm:

**Comment by hongkailiu** [comment]
/hold

@evakhoni 

I think it is ready for a pre-merge test. Thanks.

**Review by petr-muller — APPROVED** [review]
Nice job here, this was a nasty bug :+1:

**Comment by petr-muller** [comment]
/hold cancel

We do not need to hold for QE, CVO blocks merges on QE labels

**Comment by evakhoni** [comment]
pre-merge verified. 
/label qe-approved

**Comment by petr-muller** [comment]
TP jobs are busted right now because of OLM (revert is https://github.com/openshift/operator-framework-catalogd/pull/67) , consistent with our results:

```
{Operator unavailable (CatalogdDeploymentCatalogdControllerManager_Deploying): CatalogdDeploymentCatalogdControllerManagerAvailable: Waiting for Deployment  Operator unavailable (CatalogdDeploymentCatalogdControllerManager_Deploying): CatalogdDeploymentCatalogdControllerManagerAvailable: Waiting for Deployment}
```

/override ci/prow/e2e-aws-ovn-techpreview

**Pull request merged**
